File: javascript/packages/linter/src/herb-disable-comment-utils.ts
/**
 * Utilities for parsing herb:disable comments.
 *
 * Syntax:
 *   <%# herb:disable rule-name [, rule-name...] %>              # line-scoped
 *   <%# herb:disable rule-name <N> [, rule-name <N|all>...] %>  # file-scoped
 *
 * Each comma-separated entry is either:
 *   - `rule-name` alone -> line-scoped disable (existing behavior)
 *   - `rule-name N`     -> file-scoped counter (suppress N offenses)
 *   - `rule-name all`   -> file-scoped counter (suppress every offense)
 *
 * The two forms can be mixed in the same comment. Line-scoped entries are
 * surfaced through `ruleNames`/`ruleNameDetails` (so the existing meta-rules
 * that inspect them do not need to change); file-scoped entries are surfaced
 * through `fileScopedEntries` and consumed by the linter's suppression path.
 */

/**
 * Information about a single rule name in a herb:disable comment.
 */
export interface HerbDisableRuleName {
  /** The rule name */
  name: string
  /** The starting offset of this rule name within the content/line */
  offset: number
  /** The length of the rule name */
  length: number
}

/**
 * A file-scoped herb:disable entry (rule-name with a trailing count or `all`).
 */
export interface HerbDisableFileScopedEntry {
  /** The rule name */
  name: string
  /** Zero-based offset of the rule name within the content/line */
  nameOffset: number
  /** Length of the rule name */
  nameLength: number
  /** Suppression count. `"all"` means "suppress every offense of this rule". */
  count: number | "all"
  /** Zero-based offset of the count token within the content/line */
  countOffset: number
  /** Length of the count token as it appears in source */
  countLength: number
  /** The raw count token (e.g. `"3"` or `"all"`) */
  countRaw: string
}

/**
 * Result of parsing a herb:disable comment.
 */
export interface HerbDisableComment {
  /** The full matched string */
  match: string
  /** Line-scoped rule names (entries without a count/all suffix) */
  ruleNames: string[]
  /** Line-scoped rule name details with positions */
  ruleNameDetails: HerbDisableRuleName[]
  /** File-scoped counter entries (entries with a count or `all` suffix) */
  fileScopedEntries: HerbDisableFileScopedEntry[]
  /** The original rules string (e.g., "rule1, rule2 3, rule3 all") */
  rulesString: string
}

/**
 * Prefix for herb:disable comments
 */
const HERB_DISABLE_PREFIX = "herb:disable"

interface ParsedEntry {
  name: string
  nameOffset: number
  nameLength: number
  count?: number | "all"
  countOffset?: number
  countLength?: number
  countRaw?: string
}

function parseEntries(rulesString: string, rulesStringOffset: number): ParsedEntry[] | null {
  const rawEntries = rulesString.split(",")
  const entries: ParsedEntry[] = []

  let searchCursor = 0

  for (const rawEntry of rawEntries) {
    const entry = rawEntry.trim()
    if (entry.length === 0) return null

    // Find where this entry sits inside rulesString.
    const entryStart = rulesString.indexOf(entry, searchCursor)
    if (entryStart === -1) return null
    searchCursor = entryStart + entry.length

    const parts = entry.split(/\s+/)
    if (parts.length === 0 || parts.length > 2) return null

    const name = parts[0]
    if (name.length === 0) return null

    const nameOffset = rulesStringOffset + entryStart

    if (parts.length === 1) {
      entries.push({
        name,
        nameOffset,
        nameLength: name.length,
      })

      continue
    }

    const countRaw = parts[1]
    let count: number | "all"

    if (countRaw === "all") {
      count = "all"
    } else if (/^\d+$/.test(countRaw)) {
      const parsed = Number(countRaw)
      if (!Number.isFinite(parsed) || parsed < 0) return null
      count = parsed
    } else {
      return null
    }

    const countRelative = entry.indexOf(countRaw, name.length)
    if (countRelative === -1) return null
    const countOffset = rulesStringOffset + entryStart + countRelative

    entries.push({
      name,
      nameOffset,
      nameLength: name.length,
      count,
      countOffset,
      countLength: countRaw.length,
      countRaw,
    })
  }

  if (entries.length === 0) return null

  return entries
}

function buildResult(match: string, rulesString: string, entries: ParsedEntry[]): HerbDisableComment {
  const ruleNames: string[] = []
  const ruleNameDetails: HerbDisableRuleName[] = []
  const fileScopedEntries: HerbDisableFileScopedEntry[] = []

  for (const entry of entries) {
    if (entry.count === undefined) {
      ruleNames.push(entry.name)
      ruleNameDetails.push({
        name: entry.name,
        offset: entry.nameOffset,
        length: entry.nameLength,
      })
    } else {
      fileScopedEntries.push({
        name: entry.name,
        nameOffset: entry.nameOffset,
        nameLength: entry.nameLength,
        count: entry.count,
        countOffset: entry.countOffset!,
        countLength: entry.countLength!,
        countRaw: entry.countRaw!,
      })
    }
  }

  return {
    match,
    ruleNames,
    ruleNameDetails,
    fileScopedEntries,
    rulesString,
  }
}

/**
 * Parse a herb:disable comment from ERB comment content.
 * Use this when you have the content inside <%# ... %> (e.g., from ERBContentNode.content.value)
 *
 * @param content - The content string (without <%# %> delimiters)
 * @returns Parsed comment data or null if not a valid herb:disable comment
 */
export function parseHerbDisableContent(content: string): HerbDisableComment | null {
  const trimmed = content.trim()

  if (!trimmed.startsWith(HERB_DISABLE_PREFIX)) return null

  const afterPrefix = trimmed.substring(HERB_DISABLE_PREFIX.length).trimStart()
  if (afterPrefix.length === 0) return null

  const rulesString = afterPrefix.trimEnd()

  const herbDisablePrefix = content.indexOf(HERB_DISABLE_PREFIX)
  const searchStart = herbDisablePrefix + HERB_DISABLE_PREFIX.length
  const rulesStringOffset = content.indexOf(rulesString, searchStart)

  const entries = parseEntries(rulesString, rulesStringOffset)
  if (!entries) return null

  return buildResult(trimmed, rulesString, entries)
}

/**
 * Parse a herb:disable comment from a full source line.
 * Use this when you have a complete line that may contain <%# herb:disable ... %>
 *
 * @param line - The source line that may contain a herb:disable comment
 * @returns Parsed comment data or null if not a valid herb:disable comment
 */
export function parseHerbDisableLine(line: string): HerbDisableComment | null {
  const startTag = "<%#"
  const endTag = "%>"

  const startIndex = line.indexOf(startTag)
  if (startIndex === -1) return null

  const endIndex = line.indexOf(endTag, startIndex)
  if (endIndex === -1) return null

  const content = line.substring(startIndex + startTag.length, endIndex).trim()

  if (!content.startsWith(HERB_DISABLE_PREFIX)) return null

  const afterPrefix = content.substring(HERB_DISABLE_PREFIX.length).trimStart()
  if (afterPrefix.length === 0) return null

  const rulesString = afterPrefix.trimEnd()

  const herbDisablePrefix = line.indexOf(HERB_DISABLE_PREFIX)
  const searchStart = herbDisablePrefix + HERB_DISABLE_PREFIX.length
  const rulesStringOffset = line.indexOf(rulesString, searchStart)

  const entries = parseEntries(rulesString, rulesStringOffset)
  if (!entries) return null

  const fullMatch = line.substring(startIndex, endIndex + endTag.length)

  return buildResult(fullMatch, rulesString, entries)
}

/**
 * Check if an ERB comment content contains a herb:disable directive.
 *
 * @param content - The content string (without <%# %> delimiters)
 * @returns true if the content contains a herb:disable directive
 */
export function isHerbDisableContent(content: string): boolean {
  return parseHerbDisableContent(content) !== null
}

/**
 * Check if a source line contains a herb:disable comment.
 *
 * @param line - The source line
 * @returns true if the line contains a herb:disable comment
 */
export function isHerbDisableLine(line: string): boolean {
  return parseHerbDisableLine(line) !== null
}

File: rust/herb-linter/src/herb_disable_comment_utils.rs
use herb::Location;

pub struct HerbDisableRuleName {
  pub name: String,
  pub offset: usize,
  pub length: usize,
}

impl HerbDisableRuleName {
  pub fn location(&self, content_location: &Location) -> Option<Location> {
    let start_line = content_location.start.line;
    let start_column = content_location.start.column + self.offset as u32;

    Some(Location::from(start_line, start_column, start_line, start_column + self.length as u32))
  }
}

pub struct HerbDisableComment {
  pub rule_names: Vec<String>,
  pub rule_name_details: Vec<HerbDisableRuleName>,
  pub rules_string: String,
}

const HERB_DISABLE_PREFIX: &str = "herb:disable";

pub fn parse_herb_disable_content(content: &str) -> Option<HerbDisableComment> {
  let trimmed = content.trim();

  if !trimmed.starts_with(HERB_DISABLE_PREFIX) {
    return None;
  }

  let after_prefix = trimmed[HERB_DISABLE_PREFIX.len()..].trim_start();
  if after_prefix.is_empty() {
    return None;
  }

  let rules_string = after_prefix.trim_end();
  let rule_names: Vec<String> = rules_string.split(',').map(|name| name.trim().to_string()).collect();

  if rule_names.iter().any(|name| name.is_empty()) {
    return None;
  }

  if rule_names.is_empty() {
    return None;
  }

  let herb_disable_prefix_pos = content.find(HERB_DISABLE_PREFIX)?;
  let search_start = herb_disable_prefix_pos + HERB_DISABLE_PREFIX.len();
  let rules_string_offset = content[search_start..].find(rules_string)? + search_start;

  let mut rule_name_details = Vec::new();
  let mut current_offset = 0;

  for rule_name in &rule_names {
    let rule_offset = rules_string[current_offset..].find(rule_name.as_str())? + current_offset;

    rule_name_details.push(HerbDisableRuleName {
      name: rule_name.clone(),
      offset: rules_string_offset + rule_offset,
      length: rule_name.len(),
    });

    current_offset = rule_offset + rule_name.len();
  }

  Some(HerbDisableComment {
    rule_names,
    rule_name_details,
    rules_string: rules_string.to_string(),
  })
}

pub fn parse_herb_disable_line(line: &str) -> Option<HerbDisableComment> {
  let start_tag = "<%#";
  let end_tag = "%>";

  let start_index = line.find(start_tag)?;
  let end_index = line[start_index..].find(end_tag)? + start_index;
  let content = line[start_index + start_tag.len()..end_index].trim();

  if !content.starts_with(HERB_DISABLE_PREFIX) {
    return None;
  }

  let after_prefix = content[HERB_DISABLE_PREFIX.len()..].trim_start();

  if after_prefix.is_empty() {
    return None;
  }

  let rules_string = after_prefix.trim_end();
  let rule_names: Vec<String> = rules_string.split(',').map(|name| name.trim().to_string()).collect();

  if rule_names.iter().any(|name| name.is_empty()) {
    return None;
  }

  if rule_names.is_empty() {
    return None;
  }

  let herb_disable_prefix_pos = line.find(HERB_DISABLE_PREFIX)?;
  let search_start = herb_disable_prefix_pos + HERB_DISABLE_PREFIX.len();
  let rules_string_offset = line[search_start..].find(rules_string)? + search_start;

  let mut rule_name_details = Vec::new();
  let mut current_offset = 0;

  for rule_name in &rule_names {
    let rule_offset = rules_string[current_offset..].find(rule_name.as_str())? + current_offset;

    rule_name_details.push(HerbDisableRuleName {
      name: rule_name.clone(),
      offset: rules_string_offset + rule_offset,
      length: rule_name.len(),
    });

    current_offset = rule_offset + rule_name.len();
  }

  Some(HerbDisableComment {
    rule_names,
    rule_name_details,
    rules_string: rules_string.to_string(),
  })
}

#[cfg(test)]
mod tests {
  use super::*;

  #[test]
  fn test_parse_content_single_rule() {
    let result = parse_herb_disable_content(" herb:disable rule1 ").unwrap();
    assert_eq!(result.rule_names, vec!["rule1"]);
    assert_eq!(result.rules_string, "rule1");
    assert_eq!(result.rule_name_details.len(), 1);
    assert_eq!(result.rule_name_details[0].name, "rule1");
  }

  #[test]
  fn test_parse_content_multiple_rules() {
    let result = parse_herb_disable_content(" herb:disable rule1, rule2 ").unwrap();
    assert_eq!(result.rule_names, vec!["rule1", "rule2"]);
    assert_eq!(result.rules_string, "rule1, rule2");
  }

  #[test]
  fn test_parse_content_all() {
    let result = parse_herb_disable_content(" herb:disable all ").unwrap();
    assert_eq!(result.rule_names, vec!["all"]);
  }

  #[test]
  fn test_parse_content_no_rules() {
    assert!(parse_herb_disable_content(" herb:disable ").is_none());
  }

  #[test]
  fn test_parse_content_not_herb_disable() {
    assert!(parse_herb_disable_content(" some comment ").is_none());
  }

  #[test]
  fn test_parse_content_trailing_comma() {
    assert!(parse_herb_disable_content(" herb:disable rule1, ").is_none());
  }

  #[test]
  fn test_parse_content_consecutive_commas() {
    assert!(parse_herb_disable_content(" herb:disable rule1,, rule2 ").is_none());
  }

  #[test]
  fn test_parse_line_basic() {
    let result = parse_herb_disable_line("<div>test</div> <%# herb:disable rule1, rule2 %>").unwrap();
    assert_eq!(result.rule_names, vec!["rule1", "rule2"]);
  }

  #[test]
  fn test_parse_line_no_comment() {
    assert!(parse_herb_disable_line("<div>test</div>").is_none());
  }

  #[test]
  fn test_parse_line_not_herb_disable() {
    assert!(parse_herb_disable_line("<%# some comment %>").is_none());
  }
}
